File: 2021.mar.intersection-of-two-linked-lists/02.ts
// https://leetcode.com/submissions/detail/463680173/
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function getLength(node: ListNode): number {
  if (node.next === null) return 0;
  return getLength(node.next) + 1;
}

function getIntersectionNode(
  headA: ListNode | null,
  headB: ListNode | null
): ListNode | null {
  if (headA === null || headB === null) return null;
  const aLength = getLength(headA);
  const bLength = getLength(headB);

  if (aLength > bLength) {
    const diff = aLength - bLength;
    for (let i = 0; i < diff; i++) {
      headA = headA.next;
    }
  } else {
    const diff = bLength - aLength;
    for (let i = 0; i < diff; i++) {
      headB = headB.next;
    }
  }

  // console.log(headA);
  // console.log(headB);

  while (headA !== null && headA !== headB) {
    headA = headA.next;
    headB = headB.next;
  }
  return headA;
}

File: 2021.apr.palindrome-linked-list/02.rs
// https://leetcode.com/submissions/detail/475077297/
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn is_palindrome(head: Option<Box<ListNode>>) -> bool {
        let mut vec: Vec<i32> = Vec::new();
        let mut head = head;
        while let Some(h) = head {
            vec.push(h.val);
            head = h.next;
        }

        let mut cloned = vec.to_owned();
        cloned.reverse();
        vec == cloned
    }
}